sherpa/astro/utils/src/_utils.cc

```diff
@@ -96,88 +97,95 @@ namespace sherpa { namespace astro { namespace utils {
     //
     // The rmf_fold function will validate argument sizes
     //
-    
+
     npy_intp dim = npy_intp( len_counts );
     FloatArrayType counts;
     if ( EXIT_SUCCESS != counts.zeros( 1, &dim ) )
       return NULL;
-    
+
     if ( EXIT_SUCCESS != rmf_fold( source.get_size(), &source[0],
 				   num_groups.get_size(), &num_groups[0],
 				   first_chan.get_size(), &first_chan[0],
 				   num_chans.get_size(), &num_chans[0],
 				   response.get_size(), &response[0],
 				   counts.get_size(), &counts[0],
 				   npy_uintp(offset))) {
-      
+
       PyErr_SetString( PyExc_ValueError,
 		       (char*)"RMF data is invalid or inconsistent" );
       return NULL;
 
     }
 
     return counts.return_new_ref();
-    
+
   }
-      
+
   template <typename FloatArrayType, typename IntArrayType>
   PyObject* do_group( PyObject* self, PyObject* args )
   {
-    
+
     FloatArrayType data;
     FloatArrayType grouped;
     IntArrayType group;
     const char* type;
-    
+
     if ( !PyArg_ParseTuple( args, (char*)"O&O&s",
 			    (converter)convert_to_array< FloatArrayType >,
 			    &data,
 			    (converter)convert_to_array< IntArrayType >,
 			    &group,
 			    &type))
-      
+
     return NULL;
-    
+
     if ( data.get_size() != group.get_size() ) {
       ostringstream err;
       err << "input array sizes do not match, "
 	  << "data: " << data.get_size() << " vs group: " << group.get_size();
       PyErr_SetString( PyExc_TypeError, err.str().c_str() );
       return NULL;
     }
-    
-    if( EXIT_SUCCESS != _do_group(data.get_size(), data,
-				  group.get_size(), group,
-				  grouped, type) ) {
-      PyErr_SetString( PyExc_ValueError,
-		       (char*)"group data is invalid or inconsistent" );
+
+    try {
+      if( EXIT_SUCCESS != _do_group(data.get_size(), data,
+				    group.get_size(), group,
+				    grouped, type) ) {
+	PyErr_SetString( PyExc_ValueError,
+			 (char*)"group data is invalid or inconsistent" );
+	return NULL;
+      }
+    } catch ( std::out_of_range& e ) {
+      ostringstream err;
+      err << "unsupported group function: " << type;
+      PyErr_SetString( PyExc_ValueError, err.str().c_str() );
```

Comment:
The only exception thrown from the ```C++``` code is ```std::runtime_error``` so catching ```std::out_of_range``` is a No-op

```C++
(base) [dtn@devel12 sherpa]$ grep -r throw | grep -v Binary | grep -v optmethods/tests/tstoptfct.hh
optmethods/src/NelderMead.hh:          throw sherpa::OptErr( sherpa::OptErr::OutOfBound );
optmethods/src/NelderMead.hh:	throw std::runtime_error( "The reflection coefficient must be > 0" );
optmethods/src/NelderMead.hh:	throw std::runtime_error( "The expansion coefficient must be > 1" );
optmethods/src/NelderMead.hh:	throw std::runtime_error( "The contraction coefficient must be "
optmethods/src/NelderMead.hh:	throw std::runtime_error( "The shrink coefficient must be "
optmethods/src/NelderMead.hh:	throw std::runtime_error( "ERROR: Unknown contract case\n" );
optmethods/src/NelderMead.hh:	throw re;
optmethods/src/NelderMead.hh:	throw e;
optmethods/src/minpack/LevMar.hh:	  throw sherpa::OptErr( sherpa::OptErr::OutOfBound );
optmethods/src/minpack/LevMar.hh:        throw sherpa::OptErr( sherpa::OptErr::UsrFunc );
optmethods/src/minpack/LevMar.hh:        throw sherpa::OptErr( sherpa::OptErr::MaxFev );
optmethods/src/minpack/LevMar.hh:          throw sherpa::OptErr( sherpa::OptErr::OutOfBound );
optmethods/src/minpack/LevMar.hh:    //     throw sherpa::OptErr( sherpa::OptErr::UsrFunc );
optmethods/src/minpack/LevMar.hh:    //     throw sherpa::OptErr( sherpa::OptErr::MaxFev );
optmethods/src/minim.hh:	throw sherpa::OptErr( sherpa::OptErr::UsrFunc );
optmethods/src/DirectSearch.hh:	throw std::runtime_error( "Step cannot be all zeros (this would "
optmethods/src/DirectSearch.hh:	throw std::runtime_error( "The reflection coefficient must be > 0" );
optmethods/src/DirectSearch.hh:	throw std::runtime_error( "The expansion coefficient must be > 1" );
optmethods/src/DirectSearch.hh:	throw std::runtime_error( "The contraction coefficient must be "
optmethods/src/Opt.hh:	throw sherpa::OptErr( sherpa::OptErr::UsrFunc );
optmethods/src/Opt.hh:	throw sherpa::OptErr( sherpa::OptErr::MaxFev );
optmethods/src/DifEvo.hh:          throw sherpa::OptErr( sherpa::OptErr::OutOfBound );
optmethods/src/Simplex.hh:	throw std::runtime_error( "index out of bounds" );
optmethods/src/Simplex.hh:	throw std::runtime_error( "index out of bounds" );
optmethods/src/Simplex.hh:	throw std::runtime_error( "index out of bounds" );
optmethods/src/Simplex.hh:	throw std::runtime_error( "index out of bounds" );
fit.py:        # thrown by the routine, use this un-pythonic check.
ui/utils.py:            # is thrown folding a model, it will be swallowed up and the user
ui/tests/test_ui.py:provided and that an error is thrown when, for any reason, the computed covariance matrix
ui/tests/test_ui.py:    "Test an exception is thrown is the proper stat is not set"
ui/tests/test_ui.py:    "Test an exception is thrown when wstat is used without background"
ui/tests/test_ui.py:    "Test an exception is thrown if covar is not run"
estmethods/src/info_matrix.hh:		double (*fcn)(double*, int)) throw();
estmethods/src/projection.hh:				double*,double*,double*,int,int)) throw();
estmethods/src/projection.cc:				 double*,double*,double*,int,int)) throw()
estmethods/src/projection.cc:					double*,double*,double*,int,int)) throw()
estmethods/src/projection.cc:					    int)) throw()
estmethods/src/estutils.cc:	     double& answer ) throw() {
estmethods/src/estutils.cc:			 const double hardmax) throw()
estmethods/src/estutils.cc:	       const double value) throw()
estmethods/src/estutils.cc:			 const double f,const int upper) throw()
estmethods/src/estutils.cc:				      double (*fcn)(double*, int)) throw()
estmethods/src/estutils.hh:	     double& answer ) throw();
estmethods/src/estutils.hh:			 const double hardmax) throw();
estmethods/src/estutils.hh:	       const double value) throw();
estmethods/src/estutils.hh:			 const double f,const int upper) throw();
estmethods/src/estutils.hh:				      double (*fcn)(double*, int)) throw();
estmethods/src/estutils.hh:			    double (*fcn)(double*, int)) throw();
estmethods/src/estutils.hh:					    int,int)) throw();
estmethods/src/info_matrix.cc:			    double (*fcn)(double*, int)) throw()
plot/__init__.py:        # be thrown, but likely ValueError as this is raised
instrument.py:            # to be thrown here if sent directly to set_origin. So
astro/ui/tests/test_astro_session.py:    # method, so I want to double check that calc_stat_info does not throw an exception.
astro/ui/tests/test_astro_supports_pha2.py:    # some simple checks to throw in because we can
astro/ui/tests/test_serialize.py:    # Let it just throw an exception in case of failure.
astro/xspec/utils.py:    dummy function that throws an exception.
utils/tests/test_utils_error_unit.py:    standard library. The tests ensures that the decorators can still be used but if the tests are run they throw
utils/__init__.py:# derived from an exception, but is never thrown, as far as I can see.
utils/testing.py:    #  standard library. The decorator will be defined if pytest is missing, but if the tests are run they throw
utils/src/gsl/gsl_message.h:#ifdef GSL_MESSAGING_OFF        /* throw away messages */ 
stats/tests/test_user_stat.py:        pytest.fail("Call should not be throwing any exception (bug #341)")
include/sherpa/model_extension.hh:    // FIXME: throw error if ndim != 2
include/sherpa/jama.hh:	throw std::runtime_error( "singular matrix" );
include/sherpa/jama.hh:	throw std::runtime_error( "singular matrix" );
include/sherpa/myArray.hh:	throw std::runtime_error( "Operation attempted on different size arrays" );
```

Reply:
@dtnguyen2 - the error is raised by code that I added as I point out at https://github.com/sherpa/sherpa/pull/1264#pullrequestreview-744409986 (which I've marked as unresolved so the link work) - namelu

    func = funcs.at(funcname);

which I think is https://en.cppreference.com/w/cpp/container/map/at